### PR TITLE
Kmann/western

### PR DESCRIPTION
**Pull request by kylemann16**
- QoL changes
 - Clarity changes
 - Extract now creates output even if there overlapping data in tile. Data doesn't need to be consolidated before extracting.
 - Added modified filter method as new scan method 
 - Added ability to kill a task mid process, should now fail "gracefully", outputting the config to tiledb indicating that it wasn't finished, and noting which cells it touched and how many points it processed before dying. Failing gracefully works for all dask schedulers, but point updates can only be supported with a `distributed` set.

### CODE REVIEW

File: src/silvimetric/commands/scan.py [hunk 1]
```diff
@@ -1,23 +1,103 @@
...
 import logging
+import dask.bag as db
+import dask
+import math
 
-from ..resources import Storage, Data, Extents, Log
+from ..resources import Storage, Data, Extents, Bounds
 
-def scan(tdb_dir, pointcloud, bounds, point_count, resolution):
+def scan(tdb_dir, pointcloud, bounds, point_count=600000, resolution=100, depth=6):
```
Comment (hobu) [inline src/silvimetric/commands/scan.py:9]: where do these magic numbers come from?

File: src/silvimetric/resources/data.py [hunk 2]
```diff
@@ -6,6 +6,7 @@
 import pdal
 
 import pathlib
+from urllib.parse import urlparse
```
Comment (hobu) [inline src/silvimetric/resources/data.py:9]: what's this import for? appears unused.

File: src/silvimetric/resources/metric.py [hunk 3]
```diff
@@ -39,7 +40,8 @@ def to_json(self) -> dict[str, any]:
             'name': self.name,
             'dtype': np.dtype(self.dtype).str,
             'dependencies': self.dependencies,
-            'method': getsource(self._method)
+            'method_str': getsource(self._method),
+            'method': base64.b64encode(pickle.dumps(self._method)).decode()
```
Comment (hobu) [inline src/silvimetric/resources/metric.py:44]: Too much stuff happening in this one line.

I think you should maybe check if you can pickle the method before you do everything else. It's very possible that the user provided a method that doesn't pickle or is invalid. Some kind of checking on that needs to happen here before we insert it into the db.

File: src/silvimetric/resources/metric.py [hunk 4]
```diff
@@ -48,23 +50,23 @@ def from_string(data: Union[str, dict]):
         elif isinstance(data, dict):
             j = data
         name = j['name']
+        dtype = np.dtype(j['dtype'])
+        dependencies = j['dependencies']
+        method = pickle.loads(base64.b64decode(j['method'].encode()))
```
Comment (hobu) [inline src/silvimetric/resources/metric.py:55]: #doallthethings

File: src/silvimetric/resources/metric.py [hunk 5]
```diff
@@ -86,11 +88,12 @@ def m_max(data):
...
 
+#TODO change to correct dtype
 Metrics = {
-    'mean' : Metric('mean', np.float64, m_mean),
-    'mode' : Metric('mode', np.float64, m_mode),
-    'median' : Metric('median', np.float64, m_median),
-    'min' : Metric('min', np.float64, m_min),
-    'max' : Metric('max', np.float64, m_max),
-    'stddev' : Metric('stddev', np.float64, m_stddev),
+    'mean' : Metric('mean', np.float32, m_mean),
```
Comment (hobu) [inline src/silvimetric/resources/metric.py:93]: maybe the type is defaulted to np.float32. It's very rare that users would make any other type.

File: tests/test_chunk.py [hunk 6]
```diff
@@ -120,8 +121,9 @@ def test_indexing(self, extents, filtered, unfiltered):
     def test_pointcount(self, filtered, unfiltered, test_point_count, shatter_config, storage):
 
         with storage.open('w') as tdb:
-            fc = run(filtered, shatter_config, storage, tdb)
-            ufc = run(unfiltered, shatter_config, storage, tdb)
+            start_time = datetime.datetime.now().timestamp() * 1000
+            fc = run(filtered, shatter_config, storage, tdb, start_time)
```
Comment (hobu) [inline tests/test_chunk.py:125]: why thread `start_time` through all these methods instead of store it on the shatter_config?

**Review by hobu — APPROVED** [review]

File: src/silvimetric/commands/scan.py (hunk 1 above)
Comment (kylemann16) [reply src/silvimetric/commands/scan.py]: They're just magic numbers that seem to work well, and can be modified by the user if they want.

**Pull request merged**